### ISSUE TITLE
e2e: Services should be able to create a functioning NodePort service

### ISSUE THREAD

**Issue by sttts** [mesosphere/kubernetes-mesos#365]
test/e2e/service.go: 298

**Command line**: `./cluster/test-e2e.sh -v=9 -ginkgo.focus="create.a.functioning.NodePort.service"`

The test case creates a service pointing to a test webserver pod. It is tested whether it is reachable via a NodePort, i.e. the host IP and a dynamically reserved port from the service-node-port-range. For some reason the webserver pod of the testcase does not "export" a container port explicitly. Hence, the k8sm-scheduler does not reserve a HostPort from the resource offer. Hence, there is no port forwarding which can be reached by the kube-proxy of the service. Hence, the test fails.

**Test environment:** mesos/docker-compose with service-node-port-range 32000-32767 and k8sm-endpoint-controller.

**Note:** in the first test run documented at https://gist.github.com/sttts/e5fb813311bb3193357e this error did not occur because the upstream endpoint controller was used. Then direct ContainerIP addressing is used.

**Comment by sttts** [comment]
In the Kubernetes network model a port of a container is reachable without defining it in the pod spec. But then k8sm has no chance to do any port mapping magic. The only sensible solution might be to "fix" the e2e test.

**Comment by sttts** [comment]
Working example pod.yml:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  labels:
    name: nginx
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
    - containerPort: 80
```

and service.json:

```json
{
   "kind":"Service",
   "apiVersion":"v1",
   "metadata":{
      "name":"nginx",
      "labels":{
         "name":"nginx"
      }
   },
   "spec":{
      "ports": [
        {
          "port":80,
          "targetPort":80
        }
      ],
      "selector":{
         "name":"nginx"
      },
      "type": "NodePort"
   }
}
```

**Test procedure**
- Create the pod and the service.
- Then the service is reachable via 10.10.10.x:80 and via the host, e.g. 172.42.17.3:21002
- `cluster/kubectl.sh get endpoints` shows `nginx 172.42.17.3:21002`

**Counter-test procedure**
- Create the pod and the service above, **but remove the `containerPort` line from the pod.yaml.
- The service is reachable via 10.10.10.x:80
- `cluster/kubectl.sh get endpoints` shows `<none>`
- `docker inspect` on the `pause` container of the pod has not port bindings.

**Comment by sttts** [comment]
I think the e2e tests must use valid port definitions in their pods. Moreover, k8s should only proxy services to pods if the port is defined. We should try to get that into k8s.

**Comment by sttts** [comment]
A draft is available in GoogleCloudPlatform/kubernetes#10390

**Comment by sttts** [comment]
Fixed in GoogleCloudPlatform/kubernetes#10394

**Comment by sttts** [comment]
Fixed in GoogleCloudPlatform/kubernetes#10395

**Comment by sttts** [comment]
Fixed in https://github.com/kubernetes/kubernetes/pull/12917
